[찐브로스] 3주차 우테코북 기능 구현

Deocksoo opened the pull request:
3주차 코드 구현 사항입니다. 
혹시 1주차~3주차까지의 feature list를 보시려면 아래 링크를 참고해주세요!
https://github.com/ZZinBros/miniprojects-2019/wiki/feature-list

- 테스트 코드 리팩토링
  - 인수 테스트 추가
  - 코드 정리
<br>

- 친구 기능 개선
  - 양방향에서 단방향 매핑으로 변경
  - 서로 친구 요청을 해야 친구가 되도록 수정
<br>

- 게시글 공유 기능 추가
<br>

- 댓글 기능 추가
  - 댓글 추가
  - 댓글 수정
  - 댓글 삭제

andole98 committed:
refactor(post): 포스트 시간순 정렬 구현 (#170)

* feat(post): 최근에 작성된 게시글이 위쪽에 있도록 변경

* fix(post): 게시글을 수정하면 가장 위쪽에 있던 글이 수정되던 문제 해결

* refactor(post): 타임라인의 게시글 정렬기능에 전략패턴 적용

* refactor(post): 마이페이지 타임라인의 게시글에 정렬 기능 추가

* style(post): parentElement 메서드를 closest 메서드로 교체

kunggom committed:
Fix(Post): `countOfLike` 필드가 `int`로 되어 있는 문제를 수정 (#164)

* fix(Post): JPA 필드에 primitive type이 들어가 발생하는 문제점 수정

필드 정의가 Integer가 아닌 int로 되어 있었기 때문에 특정 상황에서
웹페이지 자체가 뜨지 않았습니다. 이 문제를 해결하기 위한 커밋입니다.

* fix(Post): countOfLike 필드에 기본값 추가

혹시 몰라서 아예 DB에도 기본값을 추가하도록 변경하였다.

* refactor(Post): 중복된 기본값 삭제

* fix(Post): 실수로 삭제된 부분 복구

GitHub에서 줄 단위 편집은 좀 곤란한 점이 있네요. 그냥 파일 전체를 보고 편집하는 게 나은 듯…

* fix(Post): 매뉴얼에 따라 클래스에 `@DynamicInsert` 애노테이션 추가

참고:
 * https://docs.jboss.org/hibernate/orm/5.2/userguide/html_single/Hibernate_User_Guide.html#schema-generation-column-default-value

Deocksoo committed:
refactor(front): Login area와 signup area의 중복되는 tag id 수정 (#167)

* refactor: 사용하지 않는 template 제거

1. login, signup 페이지는 entrance 페이지로 이동되어 제거합니다.
2. userpage와 관련된 템플릿들을 디렉토리를 생성하여 모두 이동시켰는데, 머지 과정에서
다시 섞여들어온 파일들이 보입니다. 정리되지 않은 해당 템플릿 파일들을 다시 삭제하였습니다.

* refactor: sign-up area와 login-area의 중복된 tag id 변경

두 템플릿을 합치는 과정에서 email과 password의 input tag id가 중복되는 버그가 있었습니다.
이를 수정합니다.

* refactor(front): entrance 페이지 왼쪽 logo 크기 수정

가입 신청서 작성 템플릿 사이즈에 맞추어 크기를 수정합니다.

yuyu154 committed:
refactor(friend): Service, Controller 클래스 리팩토링 (#169)

* test : TestBaseMock.class 생성

* refactor(test) : UserBaseTest mocking

* refactor(friend) : FriendService.class 개선

1. DTO를 매개변수로 받도록 수정

* refactor : FriendController.class 개선

* refactor : FriendService.class 에서 FriendMatcher.class 분리

kunggom committed:
fix(application.yml, PostPageController) (#172)

* fix(application.yml): 프로필 설정 파일을 여러 개로 나눔

한 yml 파일에 모든 설정을 몰아두면 설정이 서로 뒤섞이는 현상이 발견되어,
이를 해결하고자 설정 파일을 프로필 단위로 쪼갬.

* fix(PostPageController): 사라진 Sort 객체 복구

merge 과정에서 사라진 Sort 객체 생성 부분을 다시 삽입하여
컴파일 문제를 해결하였다.

yuyu154 committed:
refactor(friend): 친구 기능 개선 (단방향 요청, 친구 요청 목록) (#174)

* refactor(friend) : 단방향 요청으로 친구 기능 구현

* feat(friend) : 로그인 한 유저의 친구 요청 목록 조회 기능

* refactor(friend) : FriendControllerTest.class에 테스트 추가

* refactor(friend) : Friend 유니크 키 이름 설정

* refactor(friend) : FriendMatcher의 collect 함수 개선

andole98 committed:
test: 인수 테스트 추가 (#175)

* test(post): AcceptancePostControllerTest 생성

- 인수 테스트 추가

* refactor(common): BaseEntity에 equals() hashcode() 추가

- 하위 엔터티의 equals hashcode 삭제
- 영향받는 테스트 정리

* test: AcceptanceLoginControllerTest 생성

* test: AcceptanceLogoutControllerTest 생성

* test: AcceptanceUserControllerTest 생성

* test: AcceptanceUserEditPageControllerTest 생성

* Update src/test/java/com/woowacourse/zzinbros/post/web/PostControllerTest.java

* Update src/test/java/com/woowacourse/zzinbros/post/web/PostControllerTest.java

gch01410 committed:
refactor(like): 좋아요 기능 리팩토링 (#176)

* refactor(post): Post과 PostLike 엔티티 관계를 양방향에서 단방향으로 수정

* refactor(post): 게시글 삭제하면 좋아요도 삭제

kunggom committed:
test(DemoController): 추가 (#178)

테스트 커버리지를 늘리기 위한 추가.

yuyu154 committed:
refactor : Logout Post 방식으로 개선 (#179)

* refactor : Logout Post 방식으로 개선

* Update src/main/resources/static/js/logout.js

Co-Authored-By: JYK <40727649+andole87@users.noreply.github.com>

yuyu154 committed:
refactor(friend): Friend와 User 양방향 맵핑 (#181)

* refactor(friend) : Friend, User 객체 양방향 관계로 수정

* refactor : findFriendsByUser() 리네임

* refactor(front) : 로그아웃 상태이거나 자신의 유저 페이지일 때 친구 추가 버튼 표시 X

* refactor(user) : 친구 요청자, 친구 조회 수정

* refactor : 친구 추가, 친구 요청, 친구 완료 상태로 버튼 분리

* Update src/main/java/com/woowacourse/zzinbros/user/web/controller/FriendController.java

* Update src/main/java/com/woowacourse/zzinbros/user/web/controller/FriendController.java

* Update src/main/java/com/woowacourse/zzinbros/user/web/controller/FriendController.java

kunggom committed:
refactor(Comment): 오류 처리 및 테스트 개선 (#177)

* feat(comment): 각 댓글 사이 구분선을 제거하고 댓글 사이 간격 조절

* feat(comment): 댓글 추가 기본 기능 front 구현

댓글을 작성하였을 때 서버에 작성 요청을 날리고, 잘 작성되었다는 응답을 받으면
작성된 댓글을 화면에 추가해주도록 합니다.

* refactor(Comment): 예외 처리 개선, 테스트 보강 등

 * 예외 발생시 이를 잡는 Controller Advice 작성
 * 각 예외별로 다른 HTTP Status Code 적용
 * `CommentResponseDto` 객체를 서버측 응답에 적용
 * 댓글 작업 실패시에 대한 테스트 작성
 * 테스트 프로필에서 좀 더 자세한 로그 출력

* fix(application-test.yml): 테스트시 사용하는 H2 DB 설정 변경

테스트 때 사용하는 H2 DB에서 인 메모리 데이터베이스를 사용할 경우
원인은 모르겠지만 테스트가 정상적으로 끝나지 않는 현상이 있어서
이를 해결하기 위해 파일 기반 DB를 사용하도록 바꾸었다.

* refactor(Comment): 피드백 반영

피드백 반영 및 테스트 수정.

gch01410 committed:
feat(post): 게시글 공유기능 구현 (#182)

* feat(post-share): 게시글 공유기능 백엔드 추가

* feat(post-share): 게시글 공유횟수 백엔드 기능 추가

* feat(post-share): 게시글 공유기능 프론트 추가

* feat(post-share): 게시글 공유횟수 기능 프론트 적용

* refactor(front): 불필요한 속성 제거 및 줄 맞춤

* style(post-share): 게시글 공유 엔티티 이름 수정

* test(post-share): 게시글 공유 테스트 추가

yuyu154 committed:
feat: index페이지에 user-list 출력 기능 (#184)

* feat : index페이지에 user-list 출력 기능

* refactor : 주석 제거

andole98 committed:
refactor(friend): 친구 관계 재설정 (#185)

* refactor(friend): Friend, FriendRequest로 분리

* refactor(friend): FriendService 수정

* refactor(markup): user-page 수정

kunggom committed:
feat(Comment): 프론트엔드에서 댓글 달기, 수정, 삭제 구현 (#186)

* refacotr(front): comment.js 변수 선언 일부 수정

* feat(comment): 댓글 수정, 삭제 기능 front 구현

* refactor(Comment): 주소 체계 변경

* feat(Comment): 프론트엔드에서 댓글 수정 및 삭제가 작동하게끔 구현

Deocksoo committed:
Merge pull request #187 from ZZinBros/feat

[찐브로스] 3주차 우테코북 기능 구현

kunggom committed:
feat(logback-access): 배포 서버의 HTTP 로그를 단순하게 바꿈 (#197)

Deocksoo committed:
feat(commnet): index 페이지 comment 초기화 (#198)

* feat(comment): 기존에 달려 있는 comment도 처음에 가져오도록 수정

* refactor(comment): CommentController 메서드 컨벤션에 맞게 줄바꿈 수정

yuyu154 committed:
feat : 로그아웃 상태일 때, entrance 페이지로 이동 (#200)

kunggom committed:
feat(Comment): 댓글 단 시간을 상대 시간으로 보여주는 기능 추가 (#203)

* feat(Comment): 댓글 단 시간을 상대 시간으로 보여주는 기능 추가

여기서 사용한 프론트엔드 라이브러리와 함수를 사용하면 댓글뿐만이 아니라
Post 등 다른 곳에서도 상대 시간을 편리하게 보여줄 수 있다.

* fix: js 로드 순서 관련 변경

andole98 committed:
feat(user): 프로필 기능 추가 (#190)

* temp

* refactor(user): 유저 프로필 기능 추가

* refactor(markup): 프로필 이미지 URL 변수화

wmakerjun commented on src/main/resources/static/js/post-edit.js:
```diff
@@ -1,7 +1,7 @@
         const baseUrl = document.location.origin;
-        const contents = document.getElementById('post-edit-content');
+        const contents = event.target.closest('.modal-content').querySelector('textarea');
         const url = baseUrl + "/posts/" + contents.dataset.postid;
```
url은 [문자열 리터럴](https://poiemaweb.com/es6-template-literals)을 참고해서 표현하면 좋을 것 같아요!

wmakerjun commented on src/main/resources/static/js/signup.js:
```diff
@@ -1,6 +1,6 @@
+    const pwd = document.getElementById("signup-password");
+    const pwdConfirm = document.getElementById("signup-password-confirm");
     const inputPasswordHandler = () => {
         pwdConfirm.style.background = "#ffd5e3";
```
자바스크립트에서 style을 지정해줄 순 있지만, 가능하면 class에 style 속성을 지정하고 그 class를 추가하거나 제거하는 방법으로 구현하는게 css와 javascript의 역할을 잘 분리하는 것 같아요. background 칼라로 #ffd5e3를 넣어주는 경우 css 파일에 bg-pink와 같이 클래스명을 만들고 해당 스타일을 부여한다음, 그 클래스를 엘리먼트에 추가하거나 삭제하는 방향으로 구현해보면 어떨까요~?

yuyu154 committed:
feat : 친구 요청 거절 기능 (#205)

* refactor : logout.js 수정

* feat(friend) : 친구 요청 반환 API 구현

* feat(friend) : 친구 삭제 API

* feat(friend) : 친구 삭제 구현

* refactor(friend) : FriendServiceTest.class 추가

* feat : 친구 요청 목록은 headline에서 볼 수 있음

* refactor : 주석 제거

* refactor : TestController.class 삭제

* refactor : friend-requests-list.js 삭제

* refactor : headline에서 친구 추가할 때 새로고침

* feat(friend) : 친구 요청이 존재할 때 빨간 버튼으로 알림 기능 추가

* refactor : friend-request-btn.html layout 폴더로 이동

* feat(friend) : 친구 목록 headline에 추가

* feat : Friend 조회 API

* feat(friend) : 친구 목록 조회 (프론트)

* feat(friend) : 친구 요청 거절

* Update src/main/java/com/woowacourse/zzinbros/user/web/controller/FriendController.java

Co-Authored-By: JYK <40727649+andole87@users.noreply.github.com>

* Update src/main/resources/static/js/friend-add.js

Co-Authored-By: JYK <40727649+andole87@users.noreply.github.com>

* Update src/main/resources/static/js/friend-delete.js

Co-Authored-By: JYK <40727649+andole87@users.noreply.github.com>

* Update src/main/resources/static/js/friend-list.js

Co-Authored-By: JYK <40727649+andole87@users.noreply.github.com>

* Update src/main/resources/static/js/logout.js

Co-Authored-By: JYK <40727649+andole87@users.noreply.github.com>

* refactor(friend) : Test 통과

* refactor : logout.js 수정

* feat(friend) : 친구 요청 반환 API 구현

* feat(friend) : 친구 삭제 API

* feat(friend) : 친구 삭제 구현

* refactor(friend) : FriendServiceTest.class 추가

* feat : 친구 요청 목록은 headline에서 볼 수 있음

* refactor : 주석 제거

* refactor : TestController.class 삭제

* refactor : friend-requests-list.js 삭제

* refactor : headline에서 친구 추가할 때 새로고침

* feat(friend) : 친구 요청이 존재할 때 빨간 버튼으로 알림 기능 추가

* refactor : friend-request-btn.html layout 폴더로 이동

* feat(friend) : 친구 목록 headline에 추가

* feat : Friend 조회 API

* feat(friend) : 친구 목록 조회 (프론트)

* feat(friend) : 친구 요청 거절

* Update src/main/java/com/woowacourse/zzinbros/user/web/controller/FriendController.java

Co-Authored-By: JYK <40727649+andole87@users.noreply.github.com>

* Update src/main/resources/static/js/friend-add.js

Co-Authored-By: JYK <40727649+andole87@users.noreply.github.com>

* Update src/main/resources/static/js/friend-delete.js

Co-Authored-By: JYK <40727649+andole87@users.noreply.github.com>

* Update src/main/resources/static/js/friend-list.js

Co-Authored-By: JYK <40727649+andole87@users.noreply.github.com>

* Update src/main/resources/static/js/logout.js

Co-Authored-By: JYK <40727649+andole87@users.noreply.github.com>

* refactor(friend) : Test 통과

* refactor(test) : 컴파일 오류 해결

* debug : 테스트 코드 통과하도록 application-test.yml 수정

kunggom committed:
feat(Post): 글을 쓴 시간을 상대적으로 보여주는 기능 추가 (#206)

* 글을 쓴 시간을 `2분 전`이나 `1일 전`과 같이 상대적으로 보여주는 기능을 추가했습니다.
 * index 페이지뿐만이 아니라 모든 페이지에서 상대 시간 기능을 쓸 수 있도록 하였습니다.
 * Post 도메인 모델의 시간 속성에 대한 Getter 이름이 잘못된 것을 수정했습니다.
 * 댓글 작성 시간을 표시하는 태그를 HTML5 시맨틱에 맞게 수정하였습니다.
 * 글이나 댓글을 달았을 때 즉시 상대 시간을 표시하도록 하였습니다.
 * 시간을 표시할 때 사용할 로케일을 한국어로 고정하였습니다.
 * 상대 시간을 업데이트하는 간격을 30초로 늘렸습니다.

Deocksoo committed:
[3주차] 배포시 test 깨지는 버그 수정 (#212)

* refactor(entrance): 회원 가입했을 때 성공/실패 여부 front에서 알려주도록 수정

* refactor(test): DemoControllerTest를 인수테스트로 변경

* fix(front): index 페이지 게시하기 버튼 버그 수정

게시하기 버튼의 get 요청 url에 '/' 문자열이 하나 더 들어가는 버그를 수정하였습니다

Deocksoo committed:
refactor(login): front에서 로그인 오류 메시지 띄워주도록 수정 (#213)

* refactor(login): 로그인 실패시 front 페이지에서 오류 메시지 띄워주도록 수정

* fix(userpage): 유저페이지에서 게시글 관련 기능 동작하지 않는 버그 수정

* fix(login): 로그인 컨트롤러 실패하는 테스트 수정

yuyu154 committed:
유저 수정 페이지에서 예외에 따른 메시지 출력 (#214)

* refactor(user) : UserService.class modify() 수정

UserResponseDto를 반환하도록 수정한다

* refactor(user) : 유저 수정에 대한 API 변경

* refactor(user) : UserController.class PUT API 수정

* feat(user) : 유저 수정 페이지에서 AJAX 방식으로 이름, 이메일 수정

* refactor(user) : user-edit.js 개선

wmakerjun commented on src/main/resources/static/js/friend-list.js:
```diff
@@ -0,0 +1,84 @@
+
+        const handleDeleteFriendRequest = (event) => {
+            const requestFriendId = event.target.parentElement.value;
+            Api.delete('/friends-requests/' + requestFriendId)
```
[문자열 리터럴](https://poiemaweb.com/es6-template-literals)을 참고해서 표현하면 좋을 것 같아요!

wmakerjun commented on src/main/resources/static/js/friend-list.js:
```diff
@@ -0,0 +1,84 @@
+    const addFriend = (scrollbar, json) => {
+        const li = document.createElement('li');
+        const anchor = document.createElement('a');
+        anchor.href = '/posts?author=' + json.id;
```
[문자열 리터럴](https://poiemaweb.com/es6-template-literals)을 참고해서 표현하면 좋을 것 같아요!

wmakerjun commented on src/main/resources/static/js/social-feed.js:
```diff
@@ -1,16 +1,38 @@
+
+        let displayStrategy;
+        switch (displayStrategyText) {
+            case '전체 공개':
```
string 문자열을 직접 비교하는 것은 나중에 실수나 오류가 발생할 확률을 높이곤 해요~
따로 상태값을 담는 변수를 만들어서 IS_PUBLIC / ONLY_FRIEND / ONLY_ME 와 같은 속성에 string 값을 담아서 비교하는게 더 좋을 것 같아요. 왜냐하면 표현하는 방식이나 문구가 수정되었을 때 해당 문구를 사용하는 모든 곳을 일일히 체크할 필요가 없게 만들기 위함이에요. 

아래는 예시입니다 :)

```javascript
const DISPLAY_STARTEGY = {
	IS_PUBLIC: '전체 공개',
	ONLY_FRIEND: '친구만 공개',
	ONLY_ME: '비공개'
}
```

wmakerjun reviewed (changes requested):
자바스크립트 코드 너무 깔끔하게 잘짜서 별로 피드백 드릴게 거의 없을 정도에요!!!! 💯💯💯 
지금하던대로 남은 기간 잘 마무리해주시면 좋을 것 같습니다 :)

아주 약간의 피드백 남겼으니 해당 부분 재빨리 리팩토링하시고 다른 부분에 정진해주세요 늘 응원합니다 찐브로스 😄

wmakerjun commented on src/main/resources/static/js/comment.js:
```diff
@@ -0,0 +1,198 @@
+                let postId = post.dataset.postId;
+
+                Api.get(`/posts/${postId}/comments`)
+                    .then(res => res.json())
```
arrow function과 일반 function 표기법이 혼재되어 있습니다. 기준을 정해서 모두 통일하면 좋을 것 같아요 :)

wmakerjun commented on src/main/resources/static/js/post-share.js:
```diff
@@ -0,0 +1,23 @@
+    function onSharingClick(event) {
+        const contents = event.target.closest('.modal-content').querySelector('textarea');
+        const postId = contents.dataset.postid;
+        const url = baseUrl + "/posts/share";
```
[문자열 리터럴](https://poiemaweb.com/es6-template-literals)을 참고해서 표현하면 좋을 것 같아요!

yuyu154 replied:
백틱을 써서 표현했습니다.  
기존에 이벤트 리스너를 등록하는 부분을 버블링하는 방식으로 수정했습니다.

wmakerjun reviewed (changes requested):
너무 피드백 드린 양이 적어서 다시 보는데 공통적으로 function 선언하는 방식이 제각기인 것 같아요~ 
arrow function을 쓸 것인지 말 것인지 기준을 정해서 다같이 통일하면 좋을 것 같습니다 :)

ssosso commented on src/main/java/com/woowacourse/zzinbros/common/domain/BaseEntity.java:
```diff
@@ -19,4 +20,17 @@
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id);
```
모든 Entity가 BaseEntity에 구현된 id 기반의 hashCode를 사용해도 문제가 되지 않을까요~?

ssosso commented on src/main/java/com/woowacourse/zzinbros/user/dto/ModifyResponseMessage.java:
```diff
@@ -0,0 +1,23 @@
+    private T object;
+    private String message;
+
+    public ModifyResponseMessage(T object, String message) {
```
of 메서드가 있네요~ 생성자를 private으로 만드는 건 어떻게 생각하시나요~?

ssosso commented on src/main/java/com/woowacourse/zzinbros/user/web/controller/LogoutController.java:
```diff
@@ -17,8 +17,8 @@ public LogoutController(LoginSessionManager loginSessionManager) {
 
-    @GetMapping
-    public String logout(@SessionInfo UserSession userSessoin) {
+    @PostMapping
```
Get 이었다가 Post로 바뀐 이유가 있을까요~?

ssosso commented on src/test/java/com/woowacourse/zzinbros/post/domain/repository/PostRepositoryTest.java:
```diff
@@ -1,51 +1,53 @@
+//
+//        assertEquals(0, actual.size());
+//    }
+//}
```
좋은 테스트인데 왜 주석처리가 되었을까요~?

ssosso commented on src/main/java/com/woowacourse/zzinbros/comment/dto/CommentResponseDto.java:
```diff
@@ -0,0 +1,100 @@
+    public CommentResponseDto(final Exception exception) {
+        this.errorMessage = exception.getClass().getSimpleName() +
+                Optional.ofNullable(exception.getMessage())
+                        .map(msg -> ": " + msg).orElse("");
```
Error를 담을 때 사용하는 공통 포맷의 dto가 별도로 필요하지 않을까요~?

ssosso commented on src/main/java/com/woowacourse/zzinbros/comment/service/CommentService.java:
```diff
@@ -36,23 +48,26 @@ public Comment findById(final long commentId) {
     @Transactional
-    public Comment update(final long commentId, final String newContents, final User author) {
+    public Comment update(final CommentRequestDto requestDto, final UserSession session) {
+        final User user = userService.findLoggedInUser(session.getDto());
```
user를 service로 find하지 않고 session에 있는 id로 author 매칭을 할수는 없을까요~?

yuyu154 replied:
컨트롤러 Layer에서 확인하는 방식으로 수정했습니다!

ssosso commented on src/main/java/com/woowacourse/zzinbros/comment/service/CommentService.java:
```diff
@@ -14,18 +18,26 @@
 public class CommentService {
+    private final UserService userService;
+    private final PostService postService;
     private final CommentRepository commentRepository;
```
베디의 고민 한번 읽어보시면 좋을 것 같아요, https://velog.io/@dpudpu/4
적용하진 않으셔도 됩니다. 이런걸 고민할 수도 있구나, 고민을 해결하는 방식은 어떤게 있을 수 있을까 경험해보시면 좋을 것 같아요:)

ssosso commented on src/main/java/com/woowacourse/zzinbros/common/config/upload/UploadToFactory.java:
```diff
@@ -1,24 +1,27 @@
     }
 
     public UploadTo create(MultipartFile file) {
         if (strategy.equals("s3")) {
```
"s3", "local" 상수화되면 어떨까요~?
그리고 strategy가 점점 늘어나면 if문이 계속해서 생겨날텐데 더 좋은 방법은 없을까요~?

ssosso commented on src/main/java/com/woowacourse/zzinbros/common/config/upload/UploadToLocal.java:
```diff
@@ -20,6 +20,9 @@ public UploadToLocal(MultipartFile file,
     @Override
     public String save() {
+        if (file == null) {
+            return null;
```
null이 아닌 Optional을 반환하면 어떨까요~?

ssosso commented on src/main/java/com/woowacourse/zzinbros/post/domain/DisplayStrategy.java:
```diff
@@ -0,0 +1,25 @@
+    public static DisplayStrategy valueOf(int strategy) {
+        return Arrays.stream(values())
+                .filter(tableValue -> tableValue.match(strategy))
+                .findFirst().orElse(ALL);
```
valueOf 메서드명은 strategy에 해당하는 값이 없으면 값이 없다고 뱉어줘야 할 것 같은 느낌이 드네요;
(자바에 일반적으로 구현된 valueOf 메서드는 exception을 뱉거나 없으면 없음을 알 수 있는 값을 뱉어요)
ALL을 orElse로 준다면 메서드명에서도 이를 조금은 유추가능해야 하지 않을까요~?

ssosso replied:
그리고 @Enumerated(EnumType.STRING)을 사용하면 ALL, FRIEND, ONLY_ME와 같이 저장되는 것으로 아는데 tableValue를 별도로 저장하고 읽어와서 자동 변환해줄 수 있도록 설정하신 곳이 있을까요~?

ssosso commented on src/main/java/com/woowacourse/zzinbros/post/domain/Post.java:
```diff
@@ -49,23 +76,25 @@ public Post update(Post post) {
 
     public void addMediaFiles(MediaFile mediaFile) {
-        this.mediaFiles.add(mediaFile);
+        this.mediaFiles = mediaFile;
```
복수형 제거

ssosso commented on src/main/java/com/woowacourse/zzinbros/post/domain/SharedPost.java:
```diff
@@ -0,0 +1,44 @@
+@Entity
+public class SharedPost extends BaseEntity {
+
+    @ManyToOne
```
모든 ManyToOne에 LAZY사용 검토 부탁드립니다

ssosso commented on src/main/java/com/woowacourse/zzinbros/user/web/support/UserSession.java:
```diff
@@ -8,7 +8,7 @@
+    public UserSession(UserResponseDto loginUserDto) {
         this.loginUserDto = loginUserDto;
     }
 
```
getDto에서 새 인스턴스를 만들 때 UserResponseDto에게 copy를 요청하면 어떨까요~?

ssosso commented on src/main/java/com/woowacourse/zzinbros/user/web/controller/rest/FriendRequestController.java:
```diff
@@ -0,0 +1,38 @@
+    @DeleteMapping("/friends-requests/{id}")
+    public ResponseEntity<UserResponseDto> deleteFriendRequest(@PathVariable("id") long id,
+                                                               @SessionInfo UserSession userSession) {
+        UserResponseDto userResponseDto = userSession.getDto();
```
getDto하지 않고 Session을 그대로 Service에 넘기면 어떨까요~?
(Session 내 별도의 Dto필드가 있는 이유가 무엇인가요~?)

ssosso commented on src/main/java/com/woowacourse/zzinbros/user/domain/Friend.java:
```diff
@@ -1,74 +1,57 @@
-    public boolean isSameWithFrom(User from) {
-        return this.from.equals(from);
+    public static Friend of(User owner, User slave) {
+        return new Friend(owner, slave);
```
of를 만들어두고 new 생성자만 사용하고 있네요, 통일하면 어떨까요~?

yuyu154 replied:
new로 통일했습니다!

ssosso commented on src/main/java/com/woowacourse/zzinbros/user/domain/User.java:
```diff
@@ -28,29 +31,38 @@
     }
 
     public User(String name, @Email String email, String password) {
+        this(name, email, password, new MediaFile("/images/default/eastjun_profile.jpg"));
```
MediaFile.generateDefault 와 같은게 있으면 좋겠네요,
MediaFileService에 있는 DEFAULT_URL을 도메인으로 옮겨와서 default MediaFile 인스턴스를 생성해주는 메서드가 있으면 어떨까요~?

woowahanCU replied:
상수값을 적극적으로 사용하는 것은 어떨까요

woowahanCU commented on src/main/java/com/woowacourse/zzinbros/common/config/AppConfig.java:
```diff
@@ -19,4 +23,11 @@ public void addArgumentResolvers(List<HandlerMethodArgumentResolver> resolvers)
+    public AmazonS3 amazonS3() {
+        return AmazonS3Client.builder().standard()
+                .withRegion(Regions.AP_NORTHEAST_2)
+                .build();
```
👍 

method name은 동작을 구체적으로 드러내는 동사를 사용하시는 것이 좋다고 생각합니다.

http://www.mimul.com/pebble/default/2013/05/04/1367639300999.html

woowahanCU commented on src/main/java/com/woowacourse/zzinbros/common/config/upload/UploadFileResolver.java:
```diff
@@ -22,9 +25,16 @@ public boolean supportsParameter(MethodParameter parameter) {
-        return uploadToFactory.create(multipartFile);
+            return uploadToFactory.create(multipartFile);
+        } catch (ClassCastException e) {
+            return uploadToFactory.create(null);
```
이 메서드에서 발생할 수 있는 exception은 ClassCastException뿐인걸까요?

그리고 create 메서드에 null을 전달하는 것은 어떤 의미인걸까요?
결국 MediaFile eastjun_profile.jpg을 등록시키던데요.
우선 null로 관리하는 것이 좋을지 https://www.slideshare.net/gyumee/null-142590829
그리고 흘러가는 데이터를 기준으로 코드를 파악하게 되는데요. 객체지향적인 코드인건지
default 파일로 저 이미지를 노출하는 것은 적절한지 등등 생각해볼만한 부분이 있는거 같네요

그리고 exception을 흘려보내고 있는 부분에 대해서도 한번 생각할 부분이 있는거 같네요.
https://www.slipp.net/questions/350

woowahanCU commented on src/main/java/com/woowacourse/zzinbros/post/service/PostService.java:
```diff
@@ -55,12 +78,36 @@ public boolean delete(long postId, long userId) {
+        }
+
+        posts.addAll(postRepository.findAllByDisplayStrategyAndAuthor(DisplayStrategy.FRIEND, loginUser, sort));
+        posts.addAll(postRepository.findAllByDisplayStrategyAndAuthor(DisplayStrategy.ONLY_ME, loginUser, sort));
```
🤔
이 메서드만 놓고 봤을때 아래의 기능이 무엇인지 직관적으로 와닿지가 않네요
```java
postRepository.findAllByDisplayStrategy(DisplayStrategy.ALL, sort);
```

woowahanCU commented on src/main/java/com/woowacourse/zzinbros/post/service/PostService.java:
```diff
@@ -55,12 +78,36 @@ public boolean delete(long postId, long userId) {
-        return Collections.unmodifiableList(postRepository.findAllByAuthor(user));
+    public List<Post> readAllByUser(User user, long loginUserId, Sort sort) {
+        List<Post> posts = new ArrayList<>();
+        if (user.getId() == loginUserId) {
```
값을 꺼내와서 비교하기보다는 해당 도메인 객체에게 메시지를 전달하도록 작성해보세요.

woowahanCU commented on src/main/java/com/woowacourse/zzinbros/user/service/FriendService.java:
```diff
@@ -15,29 +17,109 @@
+
+    private void registerFriend(User sender, User receiver) {
+        friendRepository.save(new Friend(sender, receiver));
+        friendRepository.save(new Friend(receiver, sender));
```
owner_slave 의 관계가 역일 경우엔 다른 친구 관계로 읽히는데요.
이렇게 설계한 이유가 있는걸까요?

woowahanCU commented on src/main/java/com/woowacourse/zzinbros/user/web/controller/FriendController.java:
```diff
@@ -1,41 +1,52 @@
+    public ResponseEntity<Set<UserResponseDto>> getFriends(@SessionInfo UserSession userSession) {
+        final UserResponseDto loginUserDto = userSession.getDto();
+        Set<UserResponseDto> friends = friendService.findFriendsByUser(loginUserDto);
+        return new ResponseEntity<>(friends, HttpStatus.OK);
```
```suggestion
        return ResponseEntity.ok(friends);
```

https://www.baeldung.com/java-constructors-vs-static-factory-methods

woowahanCU commented on src/test/java/com/woowacourse/zzinbros/post/domain/repository/PostRepositoryTest.java:
```diff
@@ -1,51 +1,53 @@
+//        List<Post> actual = postRepository.findAllByAuthor(savedUser, by(Direction.DESC, "createdDateTime"));
+//
+//        assertEquals(0, actual.size());
+//    }
```
주석은 제거해주세요.

woowahanCU commented on src/test/java/com/woowacourse/zzinbros/post/service/PostServiceTest.java:
```diff
@@ -3,70 +3,139 @@
+        given(postRepository.findAllByDisplayStrategyAndAuthor(ALL, defaultUser, sort)).willReturn(Arrays.asList());
+
+        postRequestDto = new PostRequestDto();
+        postRequestDto.setContents(DEFAULT_CONTENT);
```
초기값을 세팅하는 부분에서 중복 코드들이 많은 경향이 있네요. 메서드를 분리하여 이를 해결할 수는 없을까요.

woowahanCU reviewed (approved):
안녕하세요! 3주간 미션진행하느라 고생많았어요 💯 
몇가지 피드백 남겼으니 확인해보시고 궁금한 점 있으면 DM 남겨주세요~

ssosso commented on src/main/java/com/woowacourse/zzinbros/user/service/UserService.java:
```diff
@@ -66,11 +83,18 @@ public UserResponseDto login(UserRequestDto userRequestDto) {
 
         if (user.matchPassword(userRequestDto.getPassword())) {
-            return new UserResponseDto(user.getId(), user.getName(), user.getEmail());
+            return new UserResponseDto(user.getId(), user.getName(), user.getEmail(), user.getProfile().getUrl());
```
User를 받아서 Dto를 만들도록 하면 어떨까요~?

ssosso reviewed (changes requested):
안녕하세요, 리뷰어 쏘쏘입니다:)
피드백드리오니 확인부탁드립니다:)

andole98 committed:
feat: 3.5주차 기능 구현 (#252)

* feat(domain): Post에 DisplayStrategy 필드 추가

* feat(service): 게시글 공개범위에 맞도록 게시글 조회

* feat(front): 게시글 공개범위 설정하는 버튼 추가

* test(post): 게시물 공개범위에 맞게 조회하는 테스트 추가

* fix(post): 전체 게시글을 조회할 경우 친구만 공개하는 나의 게시글 조회 추가

* style(yml): yml파일 인덴트 4로 수정

* fix(post): 오타 수정

* fix(post): 회원페이지에서의 게시글 작성오류 해결 (#231)

* fix(post): 회원페이지에서의 게시글을 작성오류 해결

1. 회원페이지에서 게시글을 작성하면 게시글이 2개가 생성되던 오류 해결
2. 회원페이지에서 게시글을 작성하면 지정한 공개범위와 관계업이 항상 전체공개로 생성되던 오류 해결

* style(front): user-page-add.js 삭제

* fix(entrance): entrance 페이지 logo max-width 설정

css로 최대 너비 설정을 하고, entrance.css 파일로 entrance 관련 css들을 관리하도록
수정하였습니다.

* feat(comment): 다른 유저의 댓글을 수정하는 경우 front에서 alert 하도록 수정

* refactor(Time): 상대 시간 기능 개선 (#234)

* feat(Time): 서버에서 시간을 제공하는 간단한 API 추가

서버에서 현재의 절대시각을 제공하는 간단한 API 하나를 추가했습니다.
이 API는 최대한 짧은 지연시간 안에 서버의 시간 정보를 제공하도록 되어 있습니다.
클라이언트에서는 최초 접속 또는 새로고침을 할 때 이 API를 사용할 예정입니다.

* test(Time): 서버에서 시간을 제공하는 API의 테스트 추가

* 새로 추가한 시간 API를 테스트하는 코드를 추가했습니다. 이 코드에서는 서버가 반환한 시간값이 테스트 시스템의 현재 시각과 1초 이내의 차이가 난다고 가정하였습니다. 대개의 경우 0.1초 미만의 차이가 날 것입니다.
* 시간 API를 훨씬 단순하게 만들었습니다. 디버그 모드로 로깅하는 개발 환경에서 해당 메소드 첫 호출시 약 0.1초 미만의 응답 지연이 발생하며, 첫 호출 이후에는 대개 0.005초 이내로 응답하는 것을 확인할 수 있었습니다. 실제 배포 환경에서는 0.002~0.003초 정도의 지연을 기대하고 있습니다.

* refactor(Time): API 변경

- API의 경로를 변경하였습니다.
- API가 제대로 된 JSON 포맷을 반환하도록 하였습니다.
- API가 반환하는 값을 ISO 8601 UTC 포맷으로 바꾸었습니다.
- 너무 복잡한 테스트는 필요없을 것 같아, 단순하게 변경하였습니다.

* refactor(Time): 클라이언트에서 서버 시간을 받아오도록 설정

클라이언트에서 최초 접속시에 서버의 시간을 1번 받아옵니다. 이후 새로고침이 없다면 서버로부터 시간을 받아오지 않습니다. 이렇게 하여 클라이언트의 시간 정보가 이상해도 문제가 없도록 하였습니다. 만약 서버의 시간 API를 사용할 수 없는 경우에는 클라이언트의 시간 정보를 그대로 사용합니다.

* refactor(Time): 시간 라이브러리를 Day.js로 변경

- 무겁다는 평이 있는 Moment.js를 Day.js로 교체했습니다. 본체와 한국어 로케일과 필요 플러그인까지 합쳐서 10킬로바이트도 되지 않는 대단히 가벼운 시간 라이브러리입니다.
- 글이나 댓글 작성 시간이 일시적으로 "몇 초 후"로 나타나는 문제를 해결했습니다.
- 시간 업데이트 간격을 10초로 조정하였습니다.

* 닉네임을 클릭했을 때 유저 페이지로 이동 (#235)

* refactor : 유저 이름을 클릭했을 때 유저 페이지로 이동

* refactor : 공유자의 닉네임을 눌렀을 때 유저 페이지로 이동

* refactor : 댓글 작성자의 닉네임을 눌렀을 때 유저 페이지로 이동

* Update src/main/resources/static/js/comment.js

Co-Authored-By: kunggom <32009637+kunggom@users.noreply.github.com>

* fix(multipartfile): 파일 크기 초과 핸들러 추가 (#236)

- js에서 파일 크기 체크
- ExceptionHandler 추가

* refactor(style): 코드 컨벤션 통일 (#240)

* refactor(convention): comment, common 패키지 변경

* refactor(convention): mediafile 패키지

* refactor(convention): user 패키지 변경

* refactor(convention): 패키지 구조 변경

- web.controller.rest

* feat(share): 게시글 공유할 때 공개범위 설정기능 구현 (#243)

* style(post): 게시글 공유 url 수정, 게시글 공유 메서드 이름 수정

* feat(post): 게시글 공유할 때 공개범위 설정기능 추가

* feat(front): 게시글 공유 모달에 공개범위 설정메뉴 추가

* style(front): 공개범위 설정 메서드 이름 수정

* refactor(front): 공개범위 설정 드랍다운 분리

* refactor(test): 테스트에서만 사용하는 코드 제거 (#245)

* 이름으로 유저 검색 기능 추가 (#238)

* feat(search) : 검색 기능 API 구현

* feat(search) : 검색 test 페이지 구현

드롭다운 방식으로 구현

* feat(search) : index 페이지에 검색 기능 추가

* refactor(search) : 불필요한 코드 제거

* refactor(search) : 주석 제거

* refactor(search) : 스크롤 반영

* integration : merge feat branch and search branch

* refactor : User, Search 컨트롤러 분리

* refactor : UserRepository에서 로직 처리 (포함, Pageable)

* refactor : search-name.js url 변경

* refactor : User, Search ControllerTest 분리

* refactor(test) : 테스트 클래스 리팩토링

* refactor(search) : left-profile 에 검색창 이동

* refactor(user) : 최근 가입한 순으로 유저 목록 반환

* refactor(user) : `유저 목록` -> `최근 가입한 회원`

* refactor(search) : SearchService return값 변경

* refactor(OffsetDateTime): LocalDateTime을 OffsetDateTime으로 교체 (#244)

* refactor(timestamp): 모든 타임스탬프를 OffsetDateTime 형으로 바꿈

* fix(Comment): 코드 수정 중에 날아간 부분 복구

* fix(datetime): 응답 미디어타입을 다시 정상으로 돌려둠

* feat(oauth): 페이스북 로그인 기능 추가 (#246)

- 스프링 시큐리티 oauth2

* refactor(friend) : friend-list.js 백틱 리터럴 써서 수정 (#247)

* FriendService 클래스 리팩토링 (#250)

* refactor(friend) : FriendRepository에서 바로 유저를 조회 가능하도록 수정

* refactor(friend) : FriendRequestRepository에서 바로 친구 요청자를 조회 가능하도록 수정

* fix(Comment): 댓글에 HTML이나 자바스크립트가 적용되는 문제점 해결 (#251)

* refactor(timestamp): 모든 타임스탬프를 OffsetDateTime 형으로 바꿈

* fix(Comment): 코드 수정 중에 날아간 부분 복구

* fix(datetime): 응답 미디어타입을 다시 정상으로 돌려둠

* fix(Comment): 피드백 반영 - 댓글에 자바스크립트가 들어가는 문제점 해결

* typo(EscapedCharacters): 클래스 이름 변경

* Update build.gradle

Co-Authored-By: JYK <40727649+andole87@users.noreply.github.com>

* feat(notification): backend에 새로운 게시물 알림 기능 추가 (#239)

* refactor(comment): comment.js의 editMode, deleteComment 메서드 리펙토링

undefined 체크도 함께 하고, 가독성도 높일 수 있도록 수정하였습니다.

* feat(notification): PostNotification 도메인 객체 생성

* feat(notification): NotificationRepository 추가

* feat(notification):NotificationService에 알림 저장, 조회 기능 추가

특정 사건이 발생하면 사건에 해당하는 알림이 DB에 저장되고, 나중에 해당 알림을 받을
사용자가 로그인하면 조회하도록 합니다.

* feat: PostNotificationService의 notifyPostCreation 메서드 구현

PostNotification entity의 도메인 객체의 notified 필드를 notifiedUserId로 변경하여 알림받는 사람의
id만 기억하도록 변경하였습니다. User와 직접 연관관계를 갖는 경우, 해당 user를 매번 userService를 통해
찾아와햐 하는데, 알림을 받는 user의 경우 id 이외에는 사용하지 않으므로 비효율적이기 때문입니다.

* feat(notification): NotifictaionService와 PostService 연결

PostService에서 NotificationService에 의존성을 갖도록 하였습니다. 이 때 PostService의 add
테스트가 nullpoint exception으로 정상동작하지 않아서 테스트에 notificationService mock 객체를 추가하였습니다.

* refactor(notification): notificationService에서 notify후 알림 list 반환하도록 수정

NotificationType enum도 수정하였습니다.

* refactor(notification): NotificationService의 notify 메서드 리펙토링

1. save하는 메서드를 for문 밖으로 빼내 바깥에서 saveAll하도록 수정하였습니다.
2. PostService에서 notify 실패시 logger로 기록하도록 변경하였습니다.

* refactor(notification): 알림 도메인 객체의 field 변수타입 변경

notifiedUser의 id를 가지지 않고 직접 User와 연관관계를 갖도록 수정하였습니다.

* fix: rebase후 컴파일 오류 해결

Post 도메인 객체의 변경사항을 NotificationService에 반영하였습니다.

* feat(notification): notificationController 알림 조회기능 구현

* fix(post): 공유된 post에서 유저 profile이 제대로 보이도록 수정

* Update src/main/java/com/woowacourse/zzinbros/notification/web/NotificationController.java

Co-Authored-By: JYK <40727649+andole87@users.noreply.github.com>

* feat(markup): 글 작성 알림 기능 마크업 추가

andole98 committed:
refactor(oauth): Oauth 기능 삭제

- https가 없으므로.. security 삭제

kunggom committed:
fix(Comment): 댓글 갯수가 계산되지 않는 문제를 해결

andole98 committed:
Update index.css

andole98 committed:
Update search-name.js

andole98 committed:
Update index.css

andole98 committed:
Update friend-list.js

andole98 committed:
Update friend-list.js

Deocksoo committed:
Update comment.js

woowahan-pjs reviewed (approved):
축하합니다! 백엔드 Java 레벨 2의 모든 미션 완료하였어요. 🥇 
나만의 블로그부터 미니 프로젝트까지
수많은 미션을 수행하느라 대단히 고생 많으셨습니다. 💯 
좋은 개발자가 되기 위한 소중한 밑거름이 되었길 바랍니다.
이만 Merge 하겠습니다.
